src/data_structures.rs

```diff
@@ -102,6 +98,103 @@ pub struct Proof<F: PrimeField, PC: PolynomialCommitment<F>> {
     pub pc_proof: PC::BatchProof,
 }
 
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for Proof<F, PC> {
+    fn is_valid(&self) -> bool {
+        self.commitments.is_valid() &&
+            self.evaluations.is_valid() &&
+            self.pc_proof.is_valid()
```

Comment:
I expect that the checks here are just elementwise. We could do also some additional ones such as on the number of commitments and evaluations.